feat: add `Decidable`, `Fintype`, `Encodable` linters

urkud opened the pull request:
This PR ports the `decidableClassical`, `fintypeFinite` and `encodableCountable` linters from mathlib3: all three are environment linters which flag declarations that have a hypothesis `[Decidable p]` etc., but do not use the underlying assumption in this type. Restructuring the code can make the statement more general.
The implementation is in fact quite general, and can easily adapt to include further types. PRs #17518 and #17519 are two examples of further such linters.

---

Please don't fix the failures related to `HasFDerivWithinAt.pi`. Finiteness assumptions in these lemmas will be completely removed during porting to TVS.
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

urkud committed:
Merge branch 'YK-meas-fintype' into YK-dec-PRd

urkud committed:
Merge branch 'YK-dec-PRd' into YK-dec-lint

urkud committed:
chore(PartialHomeomorph): `Fintype` -> `Finite`

urkud committed:
Merge branch 'YK-homeo-pi' into YK-dec-PRd

urkud committed:
Merge branch 'YK-dec-PRd' into YK-dec-lint

urkud committed:
chore(Analysis/BoxIntegral): `Fintype` -> `Finite`

urkud committed:
chore(InnerProductSpace/Basic): drop some `DecidableEq` assumptions

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'YK-inner-prod-dec' into YK-dec-lint

urkud committed:
Merge branch 'YK-box-integral-fintype' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
chore(PowerSeries/Basic): drop a `DecidableEq` assumption

urkud committed:
chore(LinearAlgebra/Multilinear): drop a `DecidableEq` assumption

urkud committed:
chore(CategoryTheory/Sites): `Fintype` → `Finite`

Also rename `Presieve.extensive` to `Presieve.Extensive`
to follow the naming convention.

urkud committed:
chore(SimpleGraph/Finite): drop some `DecidableEq` assumptions

- don't assume `DecidableEq` in `card_edgeFinset_le_card_choose_two`;
- make `edgeFinset_deleteEdges` work with any `Fintype G.edgeSet`
  and `Fintype (G.deleteEdges s).edgeSet` instances;
- don't assume `DecidbaleEq` in theorems about `DeleteFar`.

urkud committed:
chore(SetFamily/FourFunctions): drop some unneeded `DecidableEq` assumptions

urkud committed:
chore(SimpleGraph/DegreeSum): drop some `DecidableEq` assumptions

urkud committed:
chore(SimpleGraph/IncMatrix): drop 2 `DecidableEq` assumptions

urkud committed:
Update

urkud committed:
Merge branch 'YK-incmat' into YK-dec-lint

urkud committed:
Merge branch 'YK-deg-sum' into YK-dec-lint

urkud committed:
Merge branch 'YK-fourfunc' into YK-dec-lint

urkud committed:
Merge branch 'YK-delete-far' into YK-dec-lint

urkud committed:
Merge branch 'YK-sites' into YK-dec-lint

urkud committed:
Merge branch 'YK-linalg' into YK-dec-lint

urkud committed:
Merge branch 'YK-truncfun' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
chore(Ideal/QuotientOperations): `Fintype` -> `Finite`

urkud committed:
chore(Lebesgue/EqHaar): `Fintype` -> `Finite`

urkud committed:
chore(QuadraticForm): `Fintype -> Finite`, `DecidableEq`

urkud committed:
chore(LinearAlgebra/Finsupp): `Fintype` -> `Finite`

urkud committed:
chore(FinsuppVectorSpace): golf, `Fintype` -> `Finite`

- Golf `Finset.sum_single_ite`. Should it go to another file?
- Assume `Finite` instead of `Fintype` in `equivFun_symm_stdBasis`.

urkud committed:
Fix

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Fixup

urkud committed:
Merge branch 'YK-linalg-finsupp' into YK-dec-lint

urkud committed:
Merge branch 'YK-linalg-finsupp-vs' into YK-dec-lint

urkud committed:
Merge branch 'YK-quadratic' into YK-dec-lint

urkud committed:
Merge branch 'YK-eqhaar' into YK-dec-lint

urkud committed:
Merge branch 'YK-ideal-quotient' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Check for `Inhabited` too

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

kmill committed:
feat(Lean/Expr): define `Expr.eraseProofs`

urkud committed:
chore(Computability): `Encodable`/`Fintype` -> `Countable`/`Finite`

Also golf a proof

urkud committed:
+1 theorem

urkud committed:
Shake

urkud committed:
chore(ModelTheory/DirectLimit): `Fintype` -> `Finite`, `Encodable` -> `Countable`

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'YK-model-fintype' into YK-dec-lint

urkud committed:
Merge branch 'YK-computability' into YK-dec-lint

urkud committed:
Merge remote-tracking branch 'origin/master' into YK-dec-lint

urkud committed:
chore(Algebra/DirectLimit): drop some `DecidableEq` assumptions

urkud committed:
chore(Multilinear/Basic): golf

urkud committed:
Merge branch 'YK-dir-lim' into YK-dec-lint

urkud committed:
Merge branch 'YK-mult-norm-zero' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-eraseProofs

urkud committed:
Follow suggestions in the code review

urkud committed:
Merge branch 'YK-eraseProofs' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Update

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge fix (partial)

urkud committed:
chore(Preadditive/Biproducts): fix `Decidable`/`Fintype`

* Assume `[Finite J]` instead of `[Fintype J]`
  whenever we don't need data to formulate the theorem.
* Drop `[DecidableEq]` assumptions in `biproduct.reindex`.

urkud committed:
chore(*): `Fintype` -> `Finite`, drop `Decidable`

Also, in some cases drop unneeded `Fintype` arguments.

urkud committed:
+1 file

urkud committed:
DMatrix

urkud committed:
ChartedSpace

urkud committed:
Fix merge

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Shake imports

urkud committed:
Merge branch 'YK-matrix-fintype' into YK-dec-lint

urkud committed:
Merge branch 'YK-biprod-fintype' into YK-dec-lint

urkud committed:
chore(LpSpace): cleanup `Fintype`/`Finite`

urkud committed:
chore(Squarefree): drop a `DecidableEq` assumption

Use `classical` instead

urkud committed:
Update

urkud committed:
Merge branch 'YK-fintype-lp' into YK-dec-lint

urkud committed:
Merge branch 'YK-sqf-dec' into YK-dec-lint

urkud committed:
Fix 1 more file

urkud committed:
Merge branch 'YK-fintype-lp' into YK-dec-lint

urkud committed:
chore(Associated): add `simp`s, golf, don't use `abbrev`

urkud committed:
Update

urkud committed:
Update

urkud committed:
Fix

urkud committed:
Update

urkud committed:
Update

urkud committed:
Update Mathlib/Algebra/Associated.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

urkud committed:
Change `mk` back to `abbrev`

urkud committed:
Fix

urkud committed:
Merge branch 'YK-assoc-review' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
chore(HahnSeries): fix `Fintype`/`Finite`

urkud committed:
chore(DedekindDomain): drop a `DecidableEq` assumption

urkud committed:
Fix

urkud committed:
Update

urkud committed:
Merge branch 'master' into YK-assoc-review

urkud committed:
Merge branch 'master' into YK-hahn-fintype

urkud committed:
Merge branch 'YK-hahn-fintype' into YK-dec-lint

urkud committed:
Merge branch 'YK-dedekind-dec' into YK-dec-lint

urkud committed:
Merge branch 'YK-assoc-review' into YK-dec-lint

urkud committed:
Merge branch 'YK-fintype-lp' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-fintype-lp

urkud committed:
Fix dates

urkud committed:
Merge branch 'YK-fintype-lp' into YK-dec-lint

urkud committed:
chore(DirectSum): drop a `DecidableEq` assumption

urkud committed:
chore(Module/Zlattice): fix `Fintype`/`Finite`

urkud committed:
chore(HomOrthogonal): fix `Fintype`/`Finite`

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
chore(Star/Matrix): drop `DecidableEq` assumptions

urkud committed:
chore(AffineSpace/Matrix): fix `Fintype`/`Finite`

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Fix merge

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-assoc-review

urkud committed:
chore(Matrix/ToLin): fix `Fintype`/`Finite`

urkud committed:
chore(Matrix/ToLin): fix `Fintype`/`Finite`

urkud committed:
chore(DFinsupp/Basic): drop a `DecidableEq` assumption

Golf `DFinsupp.finite_support`,
drop an unneeded `DecidableEq` assumption.

urkud committed:
chore(Finset/Basic): drop an unneeded `DecidableEq` assumption

urkud committed:
Merge branch 'master' into YK-assoc-review

urkud committed:
Merge branch 'YK-assoc-review' into YK-dec-lint

urkud committed:
Merge branch 'YK-finset-cons' into YK-dec-lint

urkud committed:
Merge branch 'YK-dfinsupp-fin' into YK-dec-lint

urkud committed:
Merge branch 'YK-mat-tolin' into YK-dec-lint

urkud committed:
chore(Algebra/DirectSum): fix `Fintype`/`Finite`

urkud committed:
chore(Algebra/DirectSum): fix `Fintype`/`Finite`

urkud committed:
Merge branch 'YK-dir-sum-fintype' into YK-mat-tolin

urkud committed:
Fix downstream theorems

urkud committed:
Fix

urkud committed:
Merge branch 'YK-mat-tolin' into YK-dec-lint

urkud committed:
chore(AkraBazzi): fix `Fintype`/`Finite`

urkud committed:
Merge branch 'YK-akra-bazzi' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge master into YK-dec-lint (using imerge)

urkud committed:
Fix

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Update

urkud committed:
Update

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Make style linter happy

grunweg commented:
FYI: #15278, #15279 and #15280 fix some easy linter errors: I don't have further fixes in the pipeline

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Fix

urkud committed:
Merge branch 'master' into YK-dec-lint

grunweg committed:
style: use \mapsto instead of =>

grunweg committed:
chore: extract core linter logic into a separate function

And document the impossibleArgs item better.

grunweg committed:
chore: split linter in three, with custom error messages

grunweg committed:
Don't lint inductive types: how many false positives will this fix?

grunweg committed:
Merge branch 'master' into YK-dec-lint

urkud commented:
We should skip constructors of inductive types as well (at least, for now).

grunweg commented:
Do you know how to do this? (You can also leave a pointer here and I'll try to add this to the linter, when I find time. This week, and probably the next, I'm also rather busy with moving places, so cannot make promises about my available Lean time.)

grunweg committed:
Also skip constructors of inductive types.

grunweg committed:
chore(Analytic/CPolynomial): nolint the finiteFintype linter

grunweg committed:
chore(CategoryTheory/Galois): nolint the finiteFintype linter

The hypothesis is required for the proof.

grunweg committed:
chore(Symmetric/FundamentalTheorem): omit unused Fintype assumption

grunweg committed:
chore(Fintype/Card): restore finiteFintype nolints,
now that the linter has been ported

grunweg committed:
chore(Order/Heyting/Regular): restore nolint finiteFintype

grunweg committed:
Speculative: remove seemingly superfluous decidability assumptions.
The files do not use classical instances, and the proofs works fine without...
is there anything I am overlooking?

grunweg committed:
chore: split out the nonemptyInhabited linter

grunweg committed:
chore(TuringMachine): nolint the inhabitedNonempty linter,
in two instances where it is violated on purpose

grunweg committed:
chore(Util/Superscript): nolint one more proof
The assumption is necessary for the proof.

grunweg committed:
SPECULATIVE chore(Tactic/GeneraliseProofs): reduce Inhabited to Nonempty
Seems to work, but need to test this.

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Partial fix

urkud committed:
Revert

urkud committed:
Update

urkud committed:
Fix

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Revert changes to `Decidable.*`

urkud committed:
Check for `Unique`

urkud committed:
Check for `Invertible` (can't be fixed until I make `IsUnit` a `class`)

urkud committed:
Update

urkud committed:
Merge branch 'YK-unique-units' into YK-dec-lint

urkud committed:
Fix

urkud committed:
Merge branch 'YK-unique-units' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Fix

grunweg committed:
chore(Logic/Basic): nolint the decidableClassical linter
on three aliasses for a declaration beginning with Decidable.
The linter itself omits declarations beginning with Decidable,
hence it should not lint such aliasses either.

grunweg committed:
Similar fix for Encodable/Basic.

grunweg committed:
chore(RamificationInertia): remove superfluous DecidableEq

grunweg committed:
chore(JacobiSum/Basic): remove superfluous DecidableEq

grunweg committed:
chore(EllipticCurve/NormalForms): remove superfluous use of DecidableEq

grunweg committed:
chore(MvPolynomial/PDeriv): remove superfluous use of DecidableEq

grunweg committed:
chore(FunProp/RefinedDiscTree): reduce Inhabited -> Nonempty

grunweg committed:
Re-enable one nolint annotation.

grunweg committed:
fixup: PDeriv change

grunweg committed:
fixup(DomMulAct): remove decidability assumptions

grunweg committed:
One more nolint; omit partial declarations.

grunweg committed:
chore: move file to Tactic/Linter/UnusedAssumptionInType.lean

grunweg committed:
Remove two new linters for now: let's get the remaining ones landed first.

grunweg committed:
Nolint two last items.

grunweg committed:
Fix naming: Denumerable -> Countable

grunweg committed:
Update noshake.json

grunweg committed:
Merge branch 'master' into YK-dec-lint

grunweg committed:
Reduce diff.

grunweg commented:
!bench

grunweg commented:
(I worry if the linter implementation is too slow right now, and the linters should rather be merged... let's benchmark to find out!)

grunweg committed:
Merge branch 'master' into YK-dec-lint

kmill commented on Mathlib/GroupTheory/Perm/DomMulAct.lean:
```diff
@@ -106,6 +107,7 @@ theorem stabilizer_card [DecidableEq α] [DecidableEq ι] [Fintype ι] :
   · rfl
 
 /-- The cardinality of the set of permutations preserving a function -/
+@[nolint finiteFintype]
```
What is the reason that this nolint is necessary? The `Fintype` instance is used directly in the `∏` notation.

grunweg replied:
Good question, I'm not sure. @urkud Do you know?

urkud replied:
You don't need `[Fintype α]`, only `[Finite α]` here. In fact, I'm not sure if you need `[Finite α]` at all: if `α` is infinite, then one of the fibers is infinite, and RHS = 0. OTOH, LHS is infinite or empty, so you don't need `[Finite α]`.

grunweg replied:
@urkud Do you have a moment to try to make this change? I won't, until next week.

urkud replied:
Fixed.

grunweg replied:
Thanks!

kmill replied:
Ah, I thought it was complaining about the iota fintype instance so it looked like a bug. I didn't realize there was a `[Fintype α]` instance too. Thanks for the clarification (and for improving the lemma!)

urkud commented on Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:
```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2024 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
```
@grunweg Do you want to add yourself here?

grunweg replied:
Thanks, that sounds reasonable.
```suggestion
Authors: Yury Kudryashov, Michael Rothgang

urkud committed:
Update Mathlib/Algebra/MvPolynomial/PDeriv.lean

grunweg committed:
Extend authors.

grunweg commented:
!bench

urkud committed:
Update

urkud committed:
Merge branch 'master' into YK-dec-lint

grunweg commented:
!bench

grunweg committed:
Comment about partial definitions.

grunweg committed:
RFC: revert changes to GeneraliseProofs

grunweg commented:
@kmill Thanks for commenting about the `GeneralizeProofs` change; I'm learning something new every day!
The linter ignores `partial` declarations now; I just verified that the change *could* be reverted. Would you prefer that change?

grunweg committed:
Also omit recursors.

urkud commented:
@grunweg Linter flags `GeneralizeProofs` now. I can't find the relevant @kmill's comment. Was it here or on Zulip?

grunweg commented:
It was [on a dependent PR](https://github.com/leanprover-community/mathlib4/pull/17125#issuecomment-2405497324).

grunweg commented:
I'm also open to reverting [this commit](https://github.com/leanprover-community/mathlib4/pull/10235/commits/4b1e329abaf49f0fcf8aa6fab48096effdaf9034).

urkud committed:
Merge branch 'master' into YK-dec-lint

grunweg committed:
Revert second change to GeneralizeProofs: that is a linter false positive.

urkud commented:
Is there a way to test for a function being partial? If yes, then we should do it.

kmill commented:
`partial` definitions create `opaque` definitions, so it could check for that (though theoretically there could be intentional `opaque` definitions in the library, but this has not been observed yet).

Here are some checks I found:

```lean
partial def foo (n : Nat) : Nat := foo n

open Lean

def isOpaque (n : Name) : MetaM Bool := do
  let ci ← getConstInfo n
  return ci matches .opaqueInfo ..

def isUnsafe (n : Name) : MetaM Bool := do
  let ci ← getConstInfo n
  return ci.isUnsafe

def isPartial (n : Name) : MetaM Bool := do
  let ci ← getConstInfo n
  return ci.isPartial
```
With these:
* ``isOpaque `foo`` is true
* ``isUnsafe `foo`` is false (though this becomes true with the `unsafe` modifier, which also makes isOpaque become false)
* ``isPartial `foo`` is false (I think these are only used internally and no user definitions are this kind of partial)
```

grunweg committed:
Omit opaques instead.

grunweg committed:
Try reverting the generaliseProofs change again.

grunweg commented:
@kmill Thanks for the explanations. I had tried `isPartial`, but that did not do what I wanted (or did not go far enough?).
I've tried to check for `opaqueInfo` now; let's see if that works.

kmill commented:
(My cursory understanding of isPartial is that it's for "pre definitions" for (mutually) recursive functions, letting them be added to the environment for future processing. As far as I know, this option is irrelevant to us — it has no relation to `partial` — and we'd never see it unless we're really digging around into internal details.)

grunweg commented:
:bulb: That is helpful. Would you like to file PR adding this docstring? Or should I create so, so you can get help fixing it more easily :-)?

kmill commented:
Feel free to create a pr and ping me and/or Joachim. He's more familiar with this part of Lean.

grunweg commented:
!bench

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud commented:
!bench

urkud commented:
Should we merge it now?

grunweg commented:
I'm happy with merging it, and looking at post-merge benchmarking output.

grunweg commented:
I am not confident there will be no regression in linting times - but as there are further such linters in the pipeline, I am not worried about that item dropping of everybody's radar.

urkud committed:
Fix

urkud committed:
Fix more

grunweg commented:
Thanks for the fixes; they look good to me.

urkud commented:
!bench

urkud commented:
Why is it so slow?

grunweg committed:
HACK: merge them all, to test perf

grunweg committed:
Fixup

grunweg commented:
Let me test one hypothesis: does merging all linters into one help? I just pushed a commit doing so. I benchmarking is better, I would start a discussion on zulip on how to make this linter fast(er) while preserving nice error messages.

I may not be at a computer in ~40 minutes (when CI will have finished); feel free to re-bench then.

kmill commented on Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:
```diff
@@ -0,0 +1,115 @@
+        return impossibleArgs.filter fun (k, _) ↦
+          k < args.size && !e.containsFVar args[k]!.fvarId!
+  if impossibleArgs.isEmpty then return none
+  return some <| .joinSep (impossibleArgs.toList.map Prod.snd) ", "
```
I saw you were having performance problems. I tried rewriting this as efficiently as I could, but unfortunately I didn't test it, so it might be worthless... Here:
```lean
/-- Check if a given declaration assumes some hypothesis `[Type p]`, but doesn't use this
assumption in the type. `typesToAvoid` the list of such types.
This is the main logic underlying the linters below. -/
def checkUnusedAssumptionInType (declInfo : ConstantInfo) (typesToAvoid : Array Name) :
    MetaM (Option MessageData) := do
  -- We omit inductive types and their constructors, to reduce false positives.
  -- We also omit partial declarations (via the `opaque` definitions they generate):
  -- these are less useful for theorem proving, hence the linter is less useful there.
  if declInfo.isInductive || declInfo.isCtor ||
      (declInfo matches .opaqueInfo ..) then return none
  let type := declInfo.type
  -- Early return: none of the constants to avoid appear.
  unless type.containsConst typesToAvoid.contains do
    return none
  forallTelescopeReducing type fun args ty ↦ do
    -- Compute whether the given argument is used by the body of the definition outside of proofs.
    let usedIdxs : Array Bool ← do
      if ← pure declInfo.hasValue <&&> not <$> isProp type then
        lambdaTelescope declInfo.value! fun args' e ↦ do
          let mut usedIdxs := args.map (fun _ => false)
          let e ← e.eraseProofs
          let st := collectFVars {} e
          for i in [0:args'.size] do
            usedIdxs := usedIdxs.set! i (st.fvarSet.contains args'[i]!.fvarId!)
          pure usedIdxs
      else
        pure #[]
    -- Early return: every argument is used.
    if usedIdxs.size == args.size && usedIdxs.all id then
      return none
    -- Compute an array of error messages for each superfluous argument
    let ty ← ty.eraseProofs
    let mut impossibleArgs := #[]
    let mut usedFVars : CollectFVars.State := collectFVars {} ty
    for i' in [0:args.size] do
      let i := args.size - i' - 1
      let arg := args[i]!
      let t := (← inferType arg).cleanupAnnotations
      if t.containsConst typesToAvoid.contains then
        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains t.fvarId! then
          impossibleArgs := impossibleArgs.push
            (← addMessageContextFull m!"argument {i+1} {arg} : {t}")
      usedFVars := collectFVars usedFVars t
    if impossibleArgs.isEmpty then return none
    return some <| .joinSep impossibleArgs.reverse.toList ", "
```

kmill replied:
If written correctly, it should be nearly functionally identical. The only intentional difference is that the message uses `t` with cleaned-up annotations now.

grunweg replied:
Thanks! I have used this version now; let's see how it fares w.r.t. speed!

kmill commented on Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:
```diff
@@ -0,0 +1,115 @@
+  -- We omit inductive types and their constructors, to reduce false positives.
+  -- We also omit partial declarations (via the `opaque` definitions they generate):
+  -- these are less useful for theorem proving, hence the linter is less useful there.
+  if declInfo.isInductive || declInfo.isCtor ||
+      (declInfo matches .opaqueInfo ..) then return none
```
It's not a performance improvement, but you might like this syntax:
```lean4
if declInfo matches .inductInfo .. | .ctorInfo .. | .opaqueInfo .. then return none
```

grunweg replied:
I actually know this syntax - but thanks for reminding me! This started as one condition, branched out and I didn't re-visit: I will certainly use the or pattern!

urkud commented on Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:
```diff
@@ -0,0 +1,115 @@
+  if impossibleArgs.isEmpty then return none
+  return some <| .joinSep (impossibleArgs.toList.map Prod.snd) ", "
+
+@[env_linter] def allOfThem : Linter where
```
```suggestion
/-- Temporary  hack -/
@[env_linter] def allOfThem : Linter where
```

urkud committed:
Update Mathlib/Tactic/Linter/UnusedAssumptionInType.lean

urkud commented:
!bench

grunweg commented:
So it seems my theory was right: having four separate linters takes four times as long... but even the current implementation is too inefficient. Next step, try Kyle's suggestion. (I have to go bed now, but next time I work on this - or if anybody beats me to it.)

grunweg committed:
Nicer match syntax.

grunweg committed:
Small perf tweak

grunweg committed:
Untested: add Kyle's rewritten version; perf-test now!

grunweg committed:
Merge branch 'master' into YK-dec-lint

kmill commented on Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:
```diff
@@ -0,0 +1,133 @@
+      let arg := args[i]!
+      let t := (← inferType arg).cleanupAnnotations
+      if t.containsConst typesToAvoid.contains then
+        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains t.fvarId! then
```
Oops, this was supposed to be
```suggestion
        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains arg.fvarId! then
```

grunweg replied:
Thanks! Let's try the new version :-)

grunweg committed:
Fix

Co-authored-by: Kyle Miller <kmill31415@gmail.com>

grunweg committed:
Experiment: revert 8e5af956981641bf266c1dc8b65ce1ab6cb3528d.
Will this fix the build? Not a good solution, obviously!

grunweg commented:
@kmill Can you make heads or tails out of [this output](https://github.com/leanprover-community/mathlib4/actions/runs/11444480102/job/31839449455)? For comparison, I'll revert one commit, to check if there are new legitimate linter errors on master.

grunweg commented:
!bench

grunweg commented:
To clarify: this benchmarking run tests the cumulative effects of
- merging all the linters into one (which I would really prefer not to do, for sake of user-friendliness)
- adding a first early return
I expect this to still be too slow, but let's see how much :-)

grunweg commented:
Let me summarise the benchmarking results so far:
- [initial run](http://speed.lean-fro.org/mathlib4/compare/29678737-f417-40ee-bae5-eb2ec422f0a5/to/ea0e91d8-f641-406e-96eb-fa978a072335), no optimisations at all: TODO insert!
- [second run](http://speed.lean-fro.org/mathlib4/compare/29678737-f417-40ee-bae5-eb2ec422f0a5/to/ea0e91d8-f641-406e-96eb-fa978a072335): merging all linters into one, no optimisations: linting increased by +740*10⁹ instructions, wall-time by 15 seconds (in CI)
- [third run](http://speed.lean-fro.org/mathlib4/compare/4cf8ccc1-6c8f-4a8d-998a-8d741ceef6b0/to/64a43f37-fbb9-46c1-a391-0ede794462d3): merging all linters into one, no optimisations: linting increased by +150*10⁹ instructions, wall-time by ~7 seconds (in CI)

For comparison, currently, the linting step (in CI) takes 6450*10⁹ instructions or 333 seconds.

grunweg committed:
chore: un-merge the linters again, for testing

grunweg commented:
!bench

grunweg commented:
TL;DR @kmill Your help fixing the optimisation you proposed is welcome. I don't understand the code yet (and don't have much time this week to try to change that).

grunweg committed:
Remove now-superfluous import.

Command-Master commented on Mathlib/NumberTheory/JacobiSum/Basic.lean:
```diff
@@ -274,6 +274,7 @@ lemma exists_jacobiSum_eq_neg_one_add {n : ℕ} (hn : 2 < n) {χ ψ : MulChar F
     {μ : R} (hχ : χ ^ n = 1) (hψ : ψ ^ n = 1) (hn' : n ∣ Fintype.card F - 1)
     (hμ : IsPrimitiveRoot μ n) :
     ∃ z ∈ Algebra.adjoin ℤ {μ}, jacobiSum χ ψ = -1 + z * (μ - 1) ^ 2 := by
+  classical
```
What is the reason for this change?

urkud committed:
Merge master into YK-dec-lint (using imerge)

grunweg committed:
Merge branch 'master' into MR-1000-state

grunweg committed:
chore(Combinatorics): avoid DecidableEq and Fintype -> Finite

grunweg committed:
chore(Data): remove superfluous DecidableEq and Fintype

urkud committed:
Merge master into YK-dec-lint (using imerge)

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud commented:
Note: I'm working on `HasFDerivAt.pi` etc. I want to do it the right way (drop the `Normed*` assumptions at the same time), should be ready for review soon.

urkud committed:
Merge master into YK-dec-lint (using imerge)

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Fix

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

urkud committed:
Merge branch 'master' into YK-dec-lint

grunweg commented:
Can this PR be closed now, or is there anything else that's missing? If so, we could track that in #35340.
